R-package/src/xgboost_R.cc

```diff
@@ -540,15 +546,147 @@ XGB_DLL SEXP XGDMatrixNumCol_R(SEXP handle) {
   return ScalarInteger(static_cast<int>(ncol));
 }
 
+XGB_DLL SEXP XGDuplicate_R(SEXP obj) {
+  return Rf_duplicate(obj);
+}
+
+XGB_DLL SEXP XGDuplicateAttrib(SEXP src, SEXP dest) {
+  DUPLICATE_ATTRIB(dest, src);
+  return R_NilValue;
+}
+
+XGB_DLL SEXP XGPointerEqComparison(SEXP obj1, SEXP obj2) {
+  return Rf_ScalarLogical(R_ExternalPtrAddr(obj1) == R_ExternalPtrAddr(obj2));
+}
+
 // functions related to booster
-void _BoosterFinalizer(SEXP ext) {
-  if (R_ExternalPtrAddr(ext) == NULL) return;
-  CHECK_CALL(XGBoosterFree(R_ExternalPtrAddr(ext)));
-  R_ClearExternalPtr(ext);
+static void _BoosterFinalizer(SEXP R_ptr) {
+  if (R_ExternalPtrAddr(R_ptr) == NULL) return;
+  CHECK_CALL(XGBoosterFree(R_ExternalPtrAddr(R_ptr)));
+  R_ClearExternalPtr(R_ptr);
+}
+
+/* Booster is represented as an altrep list with one element which
+corresponds to an 'externalptr' holding the C object, forbidding
+modification by not implementing setters, and adding custom serialization. */
+static R_altrep_class_t XGBAltrepPointerClass;
+
+static R_xlen_t XGBAltrepPointerLength_R(SEXP R_altrepped_obj) {
+  return 1;
+}
+
+static SEXP XGBAltrepPointerGetElt_R(SEXP R_altrepped_obj, R_xlen_t idx) {
+  return R_altrep_data1(R_altrepped_obj);
+}
+
+static SEXP XGBMakeEmptyAltrep() {
+  SEXP class_name = Rf_protect(Rf_mkString("xgb.Booster"));
+  SEXP elt_names = Rf_protect(Rf_mkString("ptr"));
+  SEXP R_ptr = Rf_protect(R_MakeExternalPtr(nullptr, R_NilValue, R_NilValue));
+  SEXP R_altrepped_obj = Rf_protect(R_new_altrep(XGBAltrepPointerClass, R_ptr, R_NilValue));
+  Rf_setAttrib(R_altrepped_obj, R_NamesSymbol, elt_names);
+  Rf_setAttrib(R_altrepped_obj, R_ClassSymbol, class_name);
+  Rf_unprotect(4);
+  return R_altrepped_obj;
+}
+
+/* Note: the idea for separating this function from the one above is to be
+able to trigger all R allocations first before doing non-R allocations. */
+static void XGBAltrepSetPointer(SEXP R_altrepped_obj, BoosterHandle handle) {
+  SEXP R_ptr = R_altrep_data1(R_altrepped_obj);
+  R_SetExternalPtrAddr(R_ptr, handle);
+  R_RegisterCFinalizerEx(R_ptr, _BoosterFinalizer, TRUE);
+}
+
+const char *ubj_json_format_str = "{\"format\": \"ubj\"}";
```

Comment:
This should also be put inside an anonymous namespace, to prevent possible name collision.

Reply:
Moved to the same namespace as above.